src/bin/drag_and_drop_listbox_model.rs

```diff
@@ -0,0 +1,653 @@
+//! # ListBox and ListModel Sample
+//!
+//! This sample demonstrates how to use gtk::ListBox in combination with
+//! gio::ListStore as a model with a custom row type.
+//!
+//! It sets up a gtk::ListBox containing, per row, a label, spinbutton and
+//! an edit button. The edit button allows to edit the underlying data structure
+//! and changes are taking place immediately in the listbox by making use of GObject
+//! property bindings.
+//!
+//! In addition it is possible to add new rows and delete old ones.
+
+#[macro_use]
+extern crate glib;
+extern crate gio;
+extern crate gtk;
+
+use gio::prelude::*;
+use gtk::prelude::*;
+
+use gtk::ResponseType;
+
+use std::cell::Cell;
+use std::rc::{Rc, Weak};
+
+use std::env::args;
+
+use row_data::RowData;
+
+// make moving clones into closures more convenient
+macro_rules! clone {
+    (@param _) => ( _ );
+    (@param $x:ident) => ( $x );
+    ($($n:ident),+ => move || $body:expr) => (
+        {
+            $( let $n = $n.clone(); )+
+            move || $body
+        }
+    );
+    ($($n:ident),+ => move |$($p:tt),+| $body:expr) => (
+        {
+            $( let $n = $n.clone(); )+
+            move |$(clone!(@param $p),)+| $body
+        }
+    );
+}
+
+// upgrade weak reference or return
+#[macro_export]
+macro_rules! upgrade_weak {
+    ($x:ident, $r:expr) => {{
+        match $x.upgrade() {
+            Some(o) => o,
+            None => return $r,
+        }
+    }};
+    ($x:ident) => {
+        upgrade_weak!($x, ())
+    };
+}
+
+fn build_ui(application: &gtk::Application) {
+    // Add custom css style
+    let css_provider = gtk::CssProvider::new();
+    css_provider.load_from_data(CSS.as_bytes()).unwrap();
+    gtk::StyleContext::add_provider_for_screen(
+        gdk::Screen::get_default().as_ref().unwrap(),
+        &css_provider,
+        gtk::STYLE_PROVIDER_PRIORITY_APPLICATION,
+    );
+
+    let window = gtk::ApplicationWindow::new(application);
+
+    window.set_title("ListBox Model Sample");
+    window.set_border_width(10);
+    window.set_position(gtk::WindowPosition::Center);
+    window.set_default_size(320, 480);
+
+    let window_weak = window.downgrade();
```

Comment:
Generally it's nicer to clean/downgrade things right before the closure where they're moved. Makes it easier to understand what they are and where they come from exactly.